src/mlir_egglog/mlir_gen.py

```diff
@@ -191,81 +194,77 @@ def lookup(e):
-            return f"arith.constant {val:e} : {F32_TYPE}"
-        case ir.IntLiteral(ival=val):
-            return f"arith.constant {val} : {I32_TYPE}"
-        case ir.Symbol(name=name) if name in vars:
-            return f"%arg_{name}"
-        case ir.Symbol(name=name):
-            return f"%{name}"
-
-        # Binary Operations
-        case ir.Add(lhs=lhs, rhs=rhs):
+    # Literals and Symbols
+    match get_callable_args(expr, ir.Term.lit_f32):
+        case (f64(f),):
+            return f"arith.constant {f:e} : {F32_TYPE}"
+    match get_callable_args(expr, ir.Term.lit_i64):
+        case (i64(i),):
+            return f"arith.constant {i} : {I32_TYPE}"
+    match get_callable_args(expr, ir.Term.var):
+        case (String(var_name),):
+            return f"%arg_{var_name}" if var_name in vars else f"%{var_name}"
+    match get_callable_args(expr, ir.Term.__add__):
+        case (lhs, rhs):
             return f"arith.addf {lookup_fn(lhs)}, {lookup_fn(rhs)} : {F32_TYPE}"
-        case ir.Mul(lhs=lhs, rhs=rhs):
+    match get_callable_args(expr, ir.Term.__mul__):
+        case (lhs, rhs):
             return f"arith.mulf {lookup_fn(lhs)}, {lookup_fn(rhs)} : {F32_TYPE}"
-        case ir.Div(lhs=lhs, rhs=rhs):
+    match get_callable_args(expr, ir.Term.__truediv__):
+        case (lhs, rhs):
             return f"arith.divf {lookup_fn(lhs)}, {lookup_fn(rhs)} : {F32_TYPE}"
-        case ir.Maximum(lhs=lhs, rhs=rhs):
+    match get_callable_args(expr, ir.maximum):
+        case (lhs, rhs):
             # Maximum is handled in the generate() method for multi-operation support
             return "ERROR_MAXIMUM_HANDLED_IN_GENERATE"
-
-        # Unary Math Operations
-        case (
-            ir.Sin()
-            | ir.Cos()
-            | ir.Log()
-            | ir.Sqrt()
-            | ir.Exp()
-            | ir.Sinh()
-            | ir.Cosh()
-            | ir.Tanh()
-        ) as op:
-            op_name = type(op).__name__.lower()
-            return f"math.{op_name} {lookup_fn(op.operand)} : {F32_TYPE}"
-        case ir.Neg(operand=op):
-            return f"arith.negf {lookup_fn(op)} : {F32_TYPE}"
-
-        # Type Casting
-        case ir.CastF32(operand=op):
-            return f"arith.sitofp {lookup_fn(op)} : {I64_TYPE} to {F32_TYPE}"
-        case ir.CastI64(operand=op):
-            return f"arith.fptosi {lookup_fn(op)} : {F32_TYPE} to {I64_TYPE}"
-
-        case _:
-            raise NotImplementedError(
-                f"Unsupported expression type: {type(expr)}\n"
-                f"Expression: {expr}\n"
-                f"Supported types: FloatLiteral, IntLiteral, Symbol, Add, Mul, Div, Maximum, "
-                f"Sin, Cos, Log, Sqrt, Exp, Sinh, Cosh, Tanh, Neg, CastF32, CastI64\n"
-                f"Hint: If you're using a NumPy function, make sure it's supported by the compiler. "
-                f"Check builtin_functions.py for available operations."
-            )
+    match get_callable_args(expr, ir.sin):
```

Comment:
This control flow is a bit more verbose than the one you had above, but it does statically verify the types and arity of the functions. I am open for suggestions if you have other APIs that might do the same in a more succinct manner.